### DISCUSSION TITLE
5.1.0: new map plugin available

### DISCUSSION THREAD

**Discussion by mistic100** [mistic100/Photo-Sphere-Viewer#830] [Announcements]
A new plugin is available from version 5.1.0 : the map! It adds a minimap on the corner of the viewer that can be moved and zoomed. It also supports the addition of hotspots (compatible with the markers plugin). The map can also be maximized.

Check the documentation and demo here https://photo-sphere-viewer.js.org/plugins/map.html  
To use it right now you need `@photo-sphere-viewer/core` and `@photo-sphere-viewer/map-plugin` in version **5.1.0**

![map](https://user-images.githubusercontent.com/41597/212556193-67eadc2c-d006-4c9e-8803-e750365aa4f0.PNG)

This version is standalone but I plan to deeply integrate it with the virtual tour plugin.  
Feel free to provide you feedbacks here.

**Comment by haldo98** [comment]
author note: this comment was written for an alpha version

----

https://www.tadini.it/temp/panorama_colombe.htm

minimal configuration with a pano I already published 

I got two errors in the console for mapcomponent

I guess that's why the map does not rotate nor zoom.

**Reply by mistic100** [reply]
Thanks. I forgot to test some null values.

If you want continue testing until then :
- add `hotspots: []` to the MapPlugin configuration
- add `data: {}` to every marker

**Comment by haldo98** [comment]
author note: this comment was written for an alpha version

----

Yes Thank You.  Now works as expected.

I have two questions, just to understand how it works:

- the blue pin image is located in the lower right part of the map.   In that position the rotation of the map is smooth.and the pin center is steady while the map is rotating.
I understand that is not located in the center because You want to show a larger part of the map away from the pin.
 Is that position fixed?  Because (just a thought) if I place the map in the lower left part of the window the blue pin should be located in the lower left part of the map...   Just to mirror the whole view..
- When I maximize the window the map is placed and centered in the right part of the window. (also in your example)
Is that an expected behaviour?

Thank You Damien for your great work.

**Reply by mistic100** [reply]
I don't understand you, can you provide screenshots of what you mean ?

**Comment by haldo98** [comment]
author note: this comment was written for an alpha version

----

the attachment....
the red lines show the direction of the view in the map
the yellow lines shows the line of view from the map
![Immagine 2023-01-16 165232](https://user-images.githubusercontent.com/97630238/212724827-4a6c64b0-64c3-465b-9849-b588ff0a4688.jpg)

goes in opposite direction

if the blue pin is in the lower left part of the map ... the direction of the lines are a match

**Reply by mistic100** [reply]
This is absolutely not how it should be read! Don't try to match the point of view from the position of the map on the screen.

This is how it works (and I think **every** system with a minimap, like video games) :
- you, the viewer (the center of the photo sphere) is located on the map by the blue pin
- where you look is represented by the tip of the pin

The position of the pin is configured with `center`. The direction of the pin is corrected with `rotation` (because the map is always assumed to have the north toward the top, but the panorama might not be necessarily oriented to the north).

The map widget could be on any corner of the viewer, it would show the same thing.

---

also please respond juste bellow this message ("Write a reply") not in another comment

**Reply by mistic100** [reply]
If on init the pin is not shown on the center, this because you have a retina screen (high DPI). This was a bug and I fixed it.

**Reply by haldo98** [reply]
ok. if the pin will be on the center makes more sense... I did not understand why was in lower right part of the map.

Thank You!!!!!

**Reply by mistic100** [reply]
Now I understand what you though was displayed! It makes sense in a way but not what was planned. And much difficult to do because you can move manually the map.

I the last commit I improved the render quality on high DPI screens, it will be less blurry (especially on smartphones with density > 3)

**Reply by haldo98** [reply]
Actually I was a little bit confused when I saw the blue pin not in the center....  Thank you for the explanation.  I'll wait for the new alpha or whatever...

I tried the same pano page on my phone (xiaomi Mi11 - android 13 - chrome browser) and if i tap on the button on the pano ("+" "-" "X"  reset and maximize) nothing happen. The buttons respond (became blue for a moment)
Maybe it's the same issue.

My two cents:  Thinking about what it could be the implementation i'd love, I was thinking about having the map not shown by default.
I did not find any configuration about this.  It's not a feature request.. just to share my thoughts...

Thank You Damien.

**Reply by mistic100** [reply]
Yes the buttons are broken on alpha.1 version, it is fixed.

I will add a `visibleOnLoad` option.

**Reply by haldo98** [reply]
Great!!  and maybe also the method Hide() and show()

In my online demo 
https://www.tadini.it/temp/panorama_colombe.htm

Now I'm using a very detailed map 3625x2887 pixels  and I was thinking that maybe the zoom applied to the map in the widget should be different from the one when the map is maximized.  In my example in the widget I use 25 as default zoom, but when the map is maximized 25 is too low and I increase it so the map will cover all the window.

What do You think?

**Reply by haldo98** [reply]
also....  I added the `spotImage` configuration. I used the original svg that I pulled off from the zip-package and changed the color.
It is shown correctly in the widget but in the maximized window it disappears.

**Reply by mistic100** [reply]
I do not see either on the widget. I think this is because the SVG has no size, try to add `width="100px" height="100px"` in the file (the values don't actually matter).

This is handled automatically when configured with the SVG source directly (like on the demo, the size is extracted from viewBox), but cannot be done if the SVG is an external image.

**Reply by mistic100** [reply]
As for automatically zoom when maximizing the map, I do not want to add another param and am not sure what a good ratio would be (cannot be a constant because on a smartphone for example, the maximized view is not much larger), and it needs to be animated to not having the map jump arround when maximizing/minimizing.

---

I added "open", "close", "maximize" and "maximize" methods to the plugin.

**Reply by haldo98** [reply]
![aaaa](https://user-images.githubusercontent.com/97630238/213273176-aeae42c6-629c-439c-bf77-77c443be27d6.jpg)


it is near the border...  but it is present in the widget

**Reply by mistic100** [reply]
That must be because of the bad handling of high DPI on version alpha.1  
But I confirm you need to add a defined size in the SVG, I just tested it locally with the latests fixes, it doesn't work without it

**Reply by mistic100** [reply]
@haldo98 You can switch to version 5.1.0-alpha.2 it should fix most of the issues.

Thank you for taking time to test this.

**Reply by haldo98** [reply]
> That must be because of the bad handling of high DPI on version alpha.1 But I confirm you need to add a defined size in the SVG, I just tested it locally with the latests fixes, it doesn't work without it

Yes it works now with      `width="100px" height="100px"`

**Reply by haldo98** [reply]
> @haldo98 You can switch to version 5.1.0-alpha.2 it should fix most of the issues.
> 
> Thank you for taking time to test this.

Thank You for your great work. I love it.

**Reply by haldo98** [reply]
I uploaded a new map - not a topographic map but a real pic taken from google earth - and I use two keys "e" and "r" for toggling between the two maps. 

Is there an another way for toggling between the two maps?  like reading which map is loaded and switch to the other? (with the same key or a button)

I noticed that when I switch to the google earth map the spot disappears. It reappears when i switch to the topographic map. The pin is always present.

I use alpha2 now.

**Reply by mistic100** [reply]
You need to track yourself with a variable what is the current map if you want ot use a single key.  
Example of navbar button which toggles between two panorama https://github.com/mistic100/Photo-Sphere-Viewer/blob/main/examples/equirectangular.html#L77-L89 this is the same concept.

---

Your two maps don't have the same size, so you need to update the markers to update the x/y properties, or use yaw+distance, it should be stable between the two images (if they have the same scale).

**Reply by haldo98** [reply]
Thank You...  I added a button using the example you provided. Thank You

For the spot I use now two images of the same size and I solved.  Works perfectly.

I changed the color of the `pinImage` to have a better contrast with the topographic map.

But I Think that with the  google earth map the original pin works better.
There's no method to change that.  Is there a way?

**Reply by mistic100** [reply]
Not really "official" *, but you can try `viewer.getPlugin('map').config.pinImage = '....';` before changing the map.

\* Comparing the the viewer which has a well supported "setOption" method

**Reply by haldo98** [reply]
Great!!!!   Works perfectly  !!!
Thank You Damien !!!
